File: stingPBS/lcs.cpp
#include<bits/stdc++.h>
using namespace std;

vector<vector<int>>dp;
    int f(string text1, string text2, int i, int j){
        if(i == -1 || j == -1) {
                // if any one opf the string has been exhausted return 0
                return 0;
            }
            
            if(dp[i][j] != -1) return dp[i][j];
            
            if(text1[i] == text2[j]) {
                return dp[i][j] = 1 + f(text1, text2, i-1, j-1);
            } else {
                return dp[i][j] = max(f(text1, text2, i-1, j), f(text1, text2, i, j-1));
            }
    }
    int longestCommonSubsequence(string text1, string text2) {
        dp.clear();
        dp.resize(1005,vector<int>(1005,-1));
        return f(text1,text2,text1.size()-1,text2.size()-1);
    }
int main(){
    // https://leetcode.com/problems/longest-common-subsequence/
    return 0;
}

File: stingPBS/lcs.java
package dynamic_programming.stingPBS;
class Solution {
    String s1= new String();
    String s2= new String();
    int dp[][] = new int [1005][1005];

    int f(int i, int j){

        if(i== -1 ||  j==-1) return 0;
        if(dp[i][j]!=-1)return dp[i][j];
        if(s1.charAt(i)==s2.charAt(j)){
            return 1+f(i-1,j-1);
        }
         return dp[i][j]=  Math.max(f(i,j-1) , f(i-1,j));
    }
    public int longestCommonSubsequence(String text1, String text2) {
        s1=text1;
        s2= text2;
        for(int i=0; i<1005; i++){
            for(int j=0; j<1005; j++){
                dp[i][j]=-1;
            }
        }
        return f(text1.length()-1, text2.length()-1);
    }
}